libbeat/autodiscover/autodiscover.go

```diff
@@ -0,0 +1,243 @@
+package autodiscover
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/cfgfile"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/mitchellh/hashstructure"
+)
+
+const debugK = "autodiscover"
+
+// TODO autodiscover providers config reload
+
+// Adapter must be implemented by the beat in order to provide Autodiscover
+type Adapter interface {
+	// TODO Hints
+
+	// CreateConfig generates a valid list of configs from the given event, the received event will have all keys defined by `StartFilter`
+	CreateConfig(bus.Event) ([]*common.Config, error)
+
+	// CheckConfig tests given config to check if it will work or not, returns errors in case it won't work
+	CheckConfig(*common.Config) error
+
+	// RunnerFactory provides runner creation by feeding valid configs
+	cfgfile.RunnerFactory
+
+	// StartFilter returns the bus filter to retrieve runner start triggering events
+	StartFilter() []string
+
+	// StopFilter returns the bus filter to retrieve runner stop triggering events
+	StopFilter() []string
+}
+
+// Autodiscover process, it takes a beat adapter and user config and runs autodiscovery process, spawning
+// new modules when any configured providers does a match
+type Autodiscover struct {
+	bus       bus.Bus
+	adapter   Adapter
+	providers []Provider
+	runners   *cfgfile.Registry
+
+	// Metadata pointers (for fast read access) store runners applied metadata, allowing
+	// updates from new events
+	meta      map[uint64]*common.MapStrPointer
+	metaMutex sync.RWMutex
+
+	startListener bus.Listener
+	stopListener  bus.Listener
+}
+
+// NewAutodiscover instantiates and returns a new Autodiscover manager
+func NewAutodiscover(name string, adapter Adapter, config *Config) (*Autodiscover, error) {
+	// Init Event bus
+	bus := bus.New(name)
+
+	// Init providers
+	var providers []Provider
+	for _, providerCfg := range config.Providers {
+		provider, err := Registry.BuildProvider(bus, providerCfg)
+		if err != nil {
+			return nil, err
+		}
+		logp.Debug(debugK, "Configured autodiscover provider: %s", provider)
+		providers = append(providers, provider)
+	}
+
+	return &Autodiscover{
+		bus:       bus,
+		adapter:   adapter,
+		runners:   cfgfile.NewRegistry(),
+		providers: providers,
+		meta:      make(map[uint64]*common.MapStrPointer),
+	}, nil
+}
+
+// Start autodiscovery process
+func (a *Autodiscover) Start() {
```

Comment:
Some idea that popped up in my head when reading this: Will it be needed to enabled / disable autodiscovery dynamically? For example on startup you want the beat to autodiscover all the services that are out there and then stop the discovery for example? Or only use autodiscovery for some time enabled through centralised config? We definitively not need that now, just thinking out loud.

Reply:
this is a good point, something like modules reload could make sense for autodiscover providers